Source/Dafny/Resolver.cs

```diff
@@ -14520,22 +14586,29 @@ public class ResolveOpts {
       public readonly bool isReveal;
       public readonly bool isPostCondition;
       public readonly bool InsideOld;
+      public readonly bool isSpecification;
```

Comment:
This fields means that we are in an expression that is always ghost. For example, invariants, assertion bodies and "by" statements, calc statements, or pre/postconditions.

The codeContext might not be ghost, for example if we are in a compiled method, the codeContext is the method and `isGhost` is false. I was originally relying on it, only to discover that it's not useful to say an expression is ghost. It works only for lemma or ghost functions.

Hence, I added this option so that the resolver can quickly infer if an expression is ghost, even before ghost inference, so that it will not bother creating two types for variables in comprehensions with a ghost constraint.
We could remove it by letting the ghost inference mechanism to automatically select the right inference, but since this mechanism is new, I am more willing to put some code (isSpecification:true) that ensures every expression in a specification position is still resolved the same.

Reply:
Would it also be possible to instead of introducing the parameter `isSpecification` to wrap the codeContext using `new CodeContextWrapper(codeContext, isGhostContext: true)` ?

Reply:
That sounds like a good idea, I did not see there was a CodeContextWrapper until now. Let me try